test/upgrade/logger.go

```diff
@@ -0,0 +1,104 @@
+/*
+Copyright 2021 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package upgrade
+
+import (
+	"errors"
+	"fmt"
+	"net/url"
+	"strings"
+	"testing"
+
+	"go.uber.org/zap"
+)
+
+const testingTScheme = "testing-t"
+
+var (
+	// ErrInvalidTestingOutputMethod when given invalid testing output method.
+	ErrInvalidTestingOutputMethod = errors.New("invalid testing output method")
+	// ErrNoTestingTRegisteredForTest when no testing.TB registered for test.
+	ErrNoTestingTRegisteredForTest = errors.New("no testing.TB registered for test")
+
+	testingTeeBees = make(map[string]testing.TB)
+)
+
+func (c Configuration) logger(tb testing.TB) (*zap.SugaredLogger, error) {
+	// TODO(mgencur): Remove when dependent repositories use LogConfig instead of Log.
+	// This is for backwards compatibility.
+	if c.Log != nil {
+		return c.Log.Sugar(), nil
+	}
+
+	testName := tb.Name()
+	testingTeeBees[testName] = tb
+	cfg := c.Config
+	outs := []string{fmt.Sprintf("%s://Log/%s", testingTScheme, testName)}
+	if len(cfg.OutputPaths) == 1 && cfg.OutputPaths[0] == "stdout" {
```

Comment:
Due to this line I have to specifically set `OutputPaths` to stdout because the `zap.NewDevelopmentConfig` by default writes to stderr for both OutputPaths and ErrorOutputPaths. If I don't override the OutputPaths everything shows twice in the console log (the logger then writes to both stderr and t.Error).
```
config := zap.NewDevelopmentConfig()
	config.OutputPaths = []string{"stdout"}
	return pkgupgrade.Configuration{T: t, LogConfig: pkgupgrade.LogConfig{Config: config}}
```

The question is: Should we use the `errOuts` if the cfg.OutputPaths is configured for stderr? ...
```
if len(cfg.OutputPaths) == 1 && cfg.OutputPaths[0] == "stderr" {
   cfg.OutputPaths = errOuts
}
```
```

Reply:
This part of code would then look like this:
```
outs := []string{fmt.Sprintf("%s://Log/%s", testingTScheme, testName)}
	errOuts := []string{fmt.Sprintf("%s://Error/%s", testingTScheme, testName)}
	if len(cfg.OutputPaths) == 1 {
		switch cfg.OutputPaths[0] {
		case "stdout":
			cfg.OutputPaths = outs
		case "stderr":
			cfg.OutputPaths = errOuts
		}
	} else {
		cfg.OutputPaths = append(cfg.OutputPaths, outs...)
	}
	if len(cfg.ErrorOutputPaths) == 1 && cfg.ErrorOutputPaths[0] == "stderr" {
		cfg.ErrorOutputPaths = errOuts
	} else {
		cfg.ErrorOutputPaths = append(cfg.ErrorOutputPaths, errOuts...)
	}
```

Reply:
Perhaps it would even make sense to treat "stderr" and "stdout" specifically and do this:
For both ErrorOutputPaths and OutputPaths: if any of the paths is "stderr" , replace it with `errOuts`, if any of the paths is "stdout", replace it with `outs`, and leave the rest as it is (no matter if we have 1 or more registered paths):
```
cfg.OutputPaths = replaceStdOutErr(testName, cfg.OutputPaths)
cfg.ErrorOutputPaths = replaceStdOutErr(testName, cfg.ErrorOutputPaths)
...
func replaceStdOutErr(testName string, paths []string) []string {
	out := fmt.Sprintf("%s://Log/%s", testingTScheme, testName)
	errOut := fmt.Sprintf("%s://Error/%s", testingTScheme, testName)
	var newPaths []string
	for _, path := range paths {
		switch path {
		case "stdout":
			newPaths = append(newPaths, out)
		case "stderr":
			newPaths = append(newPaths, errOut)
		default:
			newPaths = append(newPaths, path)
		}
	}
	return newPaths
}
```

Reply:
I think this should be fixed. I proposed such change in my 2nd review PR: https://github.com/mgencur/pkg/pull/2 (this commit https://github.com/cardil/knative-pkg/commit/f644bbbdb84760c37bb3cdd1226a57fa3f5acf06)